src/selvage/core.clj

```diff
@@ -0,0 +1,141 @@
+
+(defn emit-ln [message log-map]
+  (timbre/info :log-map log-map)
+  (stdout-emit (format "%-70s\t\t\t[CID: %s]\n" message (vis/current-cid))))
+
+(defn emit-debug-ln [message log-map]
+  (when *verbose*
+    (emit-ln message log-map)))
+
+(defn announce-flow [flow-description]
+  (emit-debug-ln (str "Running flow: " flow-description)
+                 {:flow-description flow-description
+                  :log              :flow/start}))
+
+(defn save-world-debug! [name world]
+  (swap! worlds-atom assoc name world)
+  world)
+
+(def Expression s/Any)
+(def Step [(s/one (s/enum :transition :retry) 'kind)
+           (s/one Expression 'expression)
+           (s/one s/Str 'description)])
+
+(defn timed-apply [run-function & args]
+  (let [start   (System/nanoTime)
+        ret     (apply run-function args)
+        elapsed (/ (double (- (System/nanoTime) start)) 1000000.0)]
+    [elapsed ret]))
+
+
+(defn retry-expr [retry [_kind f-expr desc]]
+  `[:retry (fn [w#] ((~retry ~f-expr) w#)) ~desc])
+
+(defn- partition-group-by [pred coll]
+  (->> coll (partition-by pred) (map #(vector (pred (first %)) %))))
+
+(defn retriable-step? [[kind _f _desc]]
+  (-> kind #{:check :query} boolean))
+
+(defn run-step [[world _] [step-type f desc]]
+  (vis/with-split-cid
+    (do
+      (emit-debug-ln (str "Running " (format "%-10s" (name step-type)) " " desc)
+                     {:log       :flow/run-step
+                      :step-type step-type
+                      :step-desc desc})
+      (let [[next-world result-desc] (f world)]
+        (save-world-debug! desc next-world)
+        (if next-world
+          [next-world result-desc]
+          (reduced [next-world result-desc]))))))
+
+(defn run-step-sequence [s0 steps]
+  (reduce run-step s0 steps))
+
+(defn steps-to-step [steps]
+  `[:sequence (fn [w#] (run-step-sequence [w# ""] (list ~@steps))) "running multiple steps"])
+
+(defn retry-sequences [retry-fn steps]
+  (->> steps
+       (partition-group-by retriable-step?)
+       (mapcat (fn [[retriable-seq? steps]]
+                 (if retriable-seq?
+                   [(retry-expr retry-fn (steps-to-step steps))]
+                   steps)))))
+
+(defmulti form->var class)
+
+(defmethod form->var Symbol [s]
+  (resolve s))
+
+(defmethod form->var ISeq [l]
+  (let [[fst snd] l]
+    (cond
+      (#{'partial 'comp} fst) (form->var snd)
+      :else                   (form->var fst))))
+
+(defmethod form->var :default [_]
+  nil)
+
+(defn is-query? [form]
+  (or (-> form form->var meta :selvage.midje.flow/query)
+      (-> form form->var meta :selvage.test.flow/query)))
+
+(defn is-transition? [form]
+  (or (and (symbol? form)
+           (-> form resolve ifn?))
+      (and (coll? form)
+           (or (let [head-form (-> form first name)]
+                 (or (= head-form "fn*")
+                     (= head-form "fn")))
+               (and (-> form first resolve meta :macro)
+                    (is-transition? (macroexpand form)))))))
```

Comment:
I'm a little afraid that this check will not catch certain cases that are valid transition functions.